File: App/Sources/Design/Theme.swift
import SwiftUI
import UIKit
import SprossKern

// MARK: - Spross design tokens
//
// Warm, playful, card-centric — poster-derived, re-grounded on the growing-box
// theme: stone-and-moss paper, clay headline, ocean and forest as the
// secondaries. No asset catalog: every color adapts to light/dark through a
// dynamic UIColor provider, over the hex pairs kern's `Palette` owns.
//
// Groups read as lowercase properties off `Theme` — `Theme.spacing.lg`,
// `Theme.colors.accent` — the shape Material 3 and the W3C design-token format
// both use, so the Android side spells every token identically.
//
// Every pairing below clears WCAG AA — 4.5:1 for text, 3:1 for controls —
// in BOTH schemes. Two rules keep it that way:
//
// 1. Accents are cut at INK strength, not fill strength. The light-mode
//    values are dark enough to read as text on paper AND on their own 14 %
//    wash (the tinted-pill pattern), which is the tightest constraint;
//    a saturated fill of the same value still reads as its hue.
// 2. Text drawn ON an accent fill uses `Theme.colors.onColor`, never `.white` —
//    dark mode's accents are pastels, where white sinks to ~1.8:1.

enum Theme {

    static let spacing = Spacing()
    static let radius = Radius()
    static let reserve = Reserve()
    static let typography = Typography()
    static let prompt = Prompt()
    static let colors = Colors()

    // MARK: Spacing (pt)

    struct Spacing {
        let xs: CGFloat = 4
        let sm: CGFloat = 8
        let md: CGFloat = 12
        let lg: CGFloat = 16
        let xl: CGFloat = 24
    }

    // MARK: Corner radius family (one family, three sizes)

    struct Radius {
        /// Hero cards (review card, completion card).
        let card: CGFloat = 28
        /// Stat tiles, chips, inline panels.
        let tile: CGFloat = 20
        /// Buttons, text fields, small controls.
        let control: CGFloat = 14
    }

    // MARK: Reserved heights (pt)

    /// A session prompt card reserves the height of its own tallest ROUTINE
    /// state, so nothing below it moves as optional content comes and goes —
    /// vertical space is the scarce axis (card, input, button and keyboard
    /// share one screen), and a button that slides under the keyboard costs
    /// more than a card with air in it. The tallest routine state, not the
    /// tallest possible one: a rare face that overflows simply grows the card,
    /// because levelling every card up to an exception buys stillness with air.
    /// The reveal is exempt — it grows the card downward and reserves nothing.
    struct Reserve {
        /// Drill prompt, shared by both drill faces: one 56 pt line of digits
        /// plus the place-value pill (141.3 pt measured), and the listening
        /// card's caption plus replay glyph plus its once-per-run silent-switch
        /// line. A gap word ("Ge l＿") is the exception that grows the card —
        /// levelling every question up to it would buy stillness with air.
        let drillCard: CGFloat = 144
        /// Review prompt (compact): the by-ear prompt's 88 pt replay target,
        /// which is the tallest a prompt side gets — a word asked by ear and
        /// the same word asked by meaning then measure alike.
        let reviewCard: CGFloat = 120
        /// A tappable tile's floor — a glyph over its label, at a size a thumb
        /// finds without aiming. The letter drill's choices and the hub's entry
        /// chips are one size, so a tile never reads as two kinds of target.
        let tile: CGFloat = 72
    }

    // MARK: Type scale — SF Rounded throughout

    struct Typography {
        let hero = Font.system(.largeTitle, design: .rounded, weight: .bold)
        let title = Font.system(.title2, design: .rounded, weight: .bold)
        let headline = Font.system(.headline, design: .rounded, weight: .semibold)
        let body = Font.system(.body, design: .rounded)
        let subheadline = Font.system(.subheadline, design: .rounded)
        let caption = Font.system(.caption, design: .rounded, weight: .medium)
        let badge = Font.system(.footnote, design: .rounded, weight: .bold)

    }

    // MARK: The question's own sizes

    /// The QUESTION itself, set as large as its card can hold.
    ///
    /// Fixed points rather than Dynamic Type: a prompt card reserves a height
    /// (`Theme.reserve.drillCard`), and a prompt that grew with the reader's setting
    /// would push the field below it off the screen that reserve exists to hold
    /// still. WHAT is asked picks the size — there is room for one numeral where
    /// there is none for a whole sentence — and every face steps down to fit
    /// rather than taking a size of its own.
    struct Prompt {
        /// A numeral the whole card is about ("1 978", "14:35").
        let digits = Font.system(size: 56, weight: .bold, design: .rounded)
        /// A picture standing where the name would: a flag that IS the question.
        let glyph = Font.system(size: 64)
        /// One word with a blank in it ("Ge l＿").
        let word = Font.system(size: 40, weight: .bold, design: .rounded)
        /// A name asked about ("Deutschland") — words run longer than numerals.
        let name = Font.system(size: 32, weight: .bold, design: .rounded)
        /// A prompt made of words, laid out like one: wrapped over lines.
        let sentence = Font.system(size: 28, weight: .bold, design: .rounded)
    }

    // MARK: Adaptive colors

    /// The box's own color table, as the SwiftUI colors it paints. The hex pairs
    /// are kern's (`design/Palette.kt`) — the app links it, so it reads the values
    /// rather than keeping a second copy that can drift.
    struct Colors {
        // Surfaces — stone paper with a moss cast, never plain white/gray.
        let background = Color(Palette.shared.background)
        let surface = Color(Palette.shared.surface)
        let surfaceTint = Color(Palette.shared.surfaceTint)
        /// Decorative hairline — card edges, the reveal divider, the ring groove.
        /// Deliberately below 3:1: the card's fill and shadow carry its boundary.
        let separator = Color(Palette.shared.separator)
        /// A line that must be SEEN — the answer field's edge is a control
        /// boundary, so it owes 3:1 where the decorative hairline does not.
        let borderStrong = Color(Palette.shared.borderStrong)

        // Text — deep forest ink instead of pure black/gray.
        let textPrimary = Color(Palette.shared.textPrimary)
        let textSecondary = Color(Palette.shared.textSecondary)
        /// Text/glyphs drawn ON a saturated accent fill (buttons, article pills).
        let onColor = Color(Palette.shared.onColor)

        // Accents — ink strength (see the header note).
        let accent = Color(Palette.shared.accent)    // clay
        let teal = Color(Palette.shared.teal)        // ocean
        let success = Color(Palette.shared.success)  // forest
        let amber = Color(Palette.shared.amber)      // ochre — a near miss, or an answer shown
        let wrong = Color(Palette.shared.wrong)      // brick — a miss
        /// The consolidated/"grown" Sprosse's own color — not `teal`, which sits too close to
        /// `der` on the hue wheel to read as anything but another blue at a badge's size.
        let grown = Color(Palette.shared.grown)      // jade

        // Article colors (poster palette).
        let der = Color(Palette.shared.der)
        let die = Color(Palette.shared.die)
        let das = Color(Palette.shared.das)
    }

    /// Gender → color. Text always carries the meaning; color only reinforces.
    ///
    /// A two-gender language folds onto the SAME two hues rather than minting its
    /// own: masculine reads der-blue, feminine die-berry. Green is German's neuter
    /// alone — a language without one simply never reaches it — and a word whose
    /// gender the box cannot name degrades to neutral, exactly as a genderless
    /// target already does. Widget and watch surfaces mirror this mapping in their
    /// own palettes. WHICH article marks which gender is not decided here:
    /// that list is `kern/model/Article.kt`.
    static func genderColor(_ gender: DLGender?) -> Color {
        switch gender {
        case .masculine: return colors.der
        case .feminine: return colors.die
        case .neuter: return colors.das
        case nil: return colors.textSecondary
        }
    }
}

/// The grammatical gender a word's article marks, as this palette reads it.
/// The Design-local twin of the box's `Gender`, so components stay kern-free.
enum DLGender {
    case masculine, feminine, neuter
}

/// An article as a card face shows it: the word itself and the gender it marks,
/// as ONE value so the two can never disagree. The screen resolves the gender
/// when it builds the face — the design system only paints it.
struct DLArticle {
    let text: String
    let gender: DLGender?

    init(_ text: String, gender: DLGender?) {
        self.text = text
        self.gender = gender
    }
}

// MARK: - Color construction

private extension UIColor {
    convenience init(hex: UInt32) {
        self.init(
            red: CGFloat((hex >> 16) & 0xFF) / 255,
            green: CGFloat((hex >> 8) & 0xFF) / 255,
            blue: CGFloat(hex & 0xFF) / 255,
            alpha: 1
        )
    }
}

extension Color {
    /// Asset-free adaptive color: hex for light mode, hex for dark mode.
    init(light: UInt32, dark: UInt32) {
        self.init(uiColor: UIColor { trait in
            trait.userInterfaceStyle == .dark ? UIColor(hex: dark) : UIColor(hex: light)
        })
    }

    /// One entry of the box's own color table, as the SwiftUI color it paints.
    /// The hex pairs are kern's (`design/Palette.kt`) — the app links it, so it
    /// reads the values rather than keeping a second copy that can drift; the
    /// `Color` type and the light/dark provider below it stay native.
    init(_ swatch: Swatch) {
        self.init(light: UInt32(bitPattern: swatch.light), dark: UInt32(bitPattern: swatch.dark))
    }
}

// MARK: - Shared modifiers & button styles

extension View {
    /// The one card shadow used everywhere.
    func dlCardShadow() -> some View {
        shadow(color: .black.opacity(0.08), radius: 16, x: 0, y: 6)
    }

    /// The one card FACE: surface fill, hairline, shadow. Every card a session
    /// puts up — vocabulary, drill prompt, listening prompt — wears this, so a
    /// screen never shows two cards cut from different cloth.
    func dlCardSurface() -> some View {
        background(
            RoundedRectangle(cornerRadius: Theme.radius.card, style: .continuous)
                .fill(Theme.colors.surface)
        )
        .overlay(
            RoundedRectangle(cornerRadius: Theme.radius.card, style: .continuous)
                .strokeBorder(Theme.colors.separator.opacity(0.6), lineWidth: 1)
        )
        .dlCardShadow()
    }
}

/// Filled terracotta primary action. Never a default gray Button.
struct DLPrimaryButtonStyle: ButtonStyle {
    var color: Color = Theme.colors.accent

    func makeBody(configuration: Configuration) -> some View {
        configuration.label
            .font(Theme.typography.headline)
            .foregroundStyle(Theme.colors.onColor)
            .padding(.vertical, Theme.spacing.lg)
            .padding(.horizontal, Theme.spacing.xl)
            .frame(minHeight: 52)
            .background(color, in: RoundedRectangle(cornerRadius: Theme.radius.control, style: .continuous))
            .opacity(configuration.isPressed ? 0.85 : 1)
            .scaleEffect(configuration.isPressed ? 0.97 : 1)
            .animation(.spring(response: 0.25, dampingFraction: 0.7), value: configuration.isPressed)
    }
}

/// Soft tinted secondary action (colored text on a translucent tint).
struct DLSoftButtonStyle: ButtonStyle {
    var color: Color = Theme.colors.accent

    func makeBody(configuration: Configuration) -> some View {
        configuration.label
            .font(Theme.typography.headline)
            .foregroundStyle(color)
            .padding(.vertical, Theme.spacing.md)
            .padding(.horizontal, Theme.spacing.lg)
            .frame(minHeight: 44)
            .background(color.opacity(0.14), in: RoundedRectangle(cornerRadius: Theme.radius.control, style: .continuous))
            .opacity(configuration.isPressed ? 0.7 : 1)
            .scaleEffect(configuration.isPressed ? 0.97 : 1)
            .animation(.spring(response: 0.25, dampingFraction: 0.7), value: configuration.isPressed)
    }
}

/// Compact icon-only action: one glyph on a round tint, sized for a thumb.
struct DLIconButtonStyle: ButtonStyle {
    var color: Color = Theme.colors.accent

    func makeBody(configuration: Configuration) -> some View {
        configuration.label
            .font(Theme.typography.headline)
            .foregroundStyle(color)
            .frame(width: 40, height: 40)
            .background(color.opacity(0.14), in: Circle())
            .opacity(configuration.isPressed ? 0.7 : 1)
            .scaleEffect(configuration.isPressed ? 0.9 : 1)
            .animation(.spring(response: 0.25, dampingFraction: 0.7), value: configuration.isPressed)
    }
}

// MARK: - Preview

#Preview("Palette") {
    ScrollView {
        VStack(alignment: .leading, spacing: Theme.spacing.lg) {
            Text("preview.tokens")
                .font(Theme.typography.hero)
                .foregroundStyle(Theme.colors.textPrimary)

            HStack(spacing: Theme.spacing.sm) {
                let articles: [DLArticle] = [
                    .init("der", gender: .masculine), .init("die", gender: .feminine),
                    .init("das", gender: .neuter), .init("el", gender: .masculine),
                    .init("la", gender: .feminine),
                ]
                ForEach(articles, id: \.text) { article in
                    Text(article.text)
                        .font(Theme.typography.badge)
                        .foregroundStyle(Theme.colors.onColor)
                        .padding(.horizontal, Theme.spacing.md)
                        .padding(.vertical, Theme.spacing.xs + 2)
                        .background(Theme.genderColor(article.gender), in: Capsule())
                }
            }

            VStack(spacing: Theme.spacing.sm) {
                swatch("Accent (Terracotta)", Theme.colors.accent)
                swatch("Teal", Theme.colors.teal)
                swatch("Success", Theme.colors.success)
                swatch("Amber (Reveal)", Theme.colors.amber)
                swatch("Surface", Theme.colors.surface)
                swatch("Surface Tint", Theme.colors.surfaceTint)
            }

            Button("common.next") {}
                .buttonStyle(DLPrimaryButtonStyle())
            Button("preview.skip") {}
                .buttonStyle(DLSoftButtonStyle(color: Theme.colors.teal))
            HStack(spacing: Theme.spacing.md) {
                Button { } label: { Image(systemName: "plus") }
                    .buttonStyle(DLIconButtonStyle())
                Button { } label: { Image(systemName: "speaker.wave.2.fill") }
                    .buttonStyle(DLIconButtonStyle(color: Theme.colors.teal))
            }
        }
        .padding(Theme.spacing.xl)
    }
    .background(Theme.colors.background)
}

private func swatch(_ name: String, _ color: Color) -> some View {
    HStack {
        RoundedRectangle(cornerRadius: Theme.radius.control, style: .continuous)
            .fill(color)
            .frame(width: 56, height: 36)
        Text(name)
            .font(Theme.typography.body)
            .foregroundStyle(Theme.colors.textPrimary)
        Spacer()
    }
}

File: android/src/main/kotlin/net/spross/app/ui/Theme.kt
package net.spross.app.ui

import androidx.compose.foundation.isSystemInDarkTheme
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Shapes
import androidx.compose.material3.Typography
import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.runtime.ReadOnlyComposable
import androidx.compose.runtime.staticCompositionLocalOf
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.ExperimentalTextApi
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.Font
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontStyle
import androidx.compose.ui.text.font.FontVariation
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.TextUnit
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.isSpecified
import androidx.compose.ui.unit.sp
import net.spross.app.R
import net.spross.kern.model.Gender
import net.spross.kern.model.Language
import net.spross.kern.model.articleGender

// Spross design tokens, Android cut: the corner family, the type ramp, the spacing and
// reserve scales, and the theme that provides them. The colors are ThemeColors.kt.

/**
 * One corner family, three sizes — control 14, tile 20, card 28.
 *
 * The extra-small slot takes the control radius too: the only thing reading it here is a
 * text field, and a text field is a control. A filled button asks for a full pill unless
 * its call site says otherwise, so those pass `shapes.small` by hand. Compose has no
 * continuous (squircle) corner; a plain rounded one is the accepted difference.
 */
val SprossShapes = Shapes(
    extraSmall = RoundedCornerShape(14.dp),
    small = RoundedCornerShape(14.dp),
    medium = RoundedCornerShape(20.dp),
    large = RoundedCornerShape(28.dp),
    extraLarge = RoundedCornerShape(28.dp),
)

/**
 * The rounded face the canonical ramp asks for, in the nearest thing Android has.
 *
 * SF Rounded is Apple's and ships with no counterpart here, so the ramp stayed on Roboto —
 * and Roboto at M3's stock tracking is what made this cut read a decade older than the iOS
 * one. Nunito carries the same voice: humanist, rounded terminals, apertures wide enough to
 * hold up at caption size. It covers Latin, Latin-ext and Cyrillic, so a Ukrainian target is
 * set in the same face as a German one, and ONE variable file (`wght` 200–1000) serves every
 * weight the ramp names. Its license is `android/licenses/Nunito-OFL.txt`, which ships in
 * the APK and folds open on the About screen.
 *
 * `♀` and `✔` fall outside its coverage and drop to the platform's symbol font. That is where
 * those two belong anyway — they are marks, not text.
 */
private val Nunito: FontFamily by lazy {
    FontFamily(
        weightedNunito(FontWeight.Normal),
        weightedNunito(FontWeight.Medium),
        weightedNunito(FontWeight.SemiBold),
        weightedNunito(FontWeight.Bold),
    )
}

/**
 * One instance of the variable file, pinned to one point on its weight axis.
 *
 * Both halves are needed and they are not the same thing: the [FontWeight] is what the ramp
 * is MATCHED on, `variationSettings` is what actually cuts the outline. Register the four
 * without the settings and every one of them renders at the file's default — ExtraLight.
 *
 * The variation-settings overload is still marked experimental; it is the only way to cut a
 * variable file from Compose, and the alternative is four static faces at three times the size.
 */
@OptIn(ExperimentalTextApi::class)
private fun weightedNunito(weight: FontWeight) =
    Font(
        R.font.nunito,
        weight,
        variationSettings = FontVariation.Settings(weight, FontStyle.Normal),
    )

/**
 * A ramp entry on the rounded face, set SOLID.
 *
 * Tracking goes to zero everywhere. M3 cuts Roboto with up to +0.5 sp, which is the single
 * loudest "stock Android" tell in the whole ramp, and the canonical ramp is set solid — the
 * iOS cut inherits SF's own near-zero tracking without asking for it.
 *
 * A null [weight] or [size] keeps whatever the M3 slot already carries. A resized slot takes
 * its leading with it: M3's line heights are cut for M3's sizes, and holding one while moving
 * the other tightens the block instead of just growing the letters.
 */
private fun TextStyle.rounded(weight: FontWeight? = null, size: TextUnit? = null) = copy(
    fontFamily = Nunito,
    fontWeight = weight ?: fontWeight,
    fontSize = size ?: fontSize,
    lineHeight = if (size == null || !lineHeight.isSpecified) {
        lineHeight
    } else {
        (lineHeight.value + (size.value - fontSize.value)).sp
    },
    letterSpacing = 0.sp,
)

/**
 * The canonical ramp's WEIGHTS and SIZES on the rounded face.
 *
 * The emphasis pattern is the iOS one: bold heroes and stat values, bold titles, semibold
 * headlines, medium captions, bold badges. Every slot is listed, including the ones that
 * keep their M3 value — a slot left out is a slot still set in Roboto.
 *
 * Each slot below names the `Theme.typography` token it answers for, and is cut to match it. M3's
 * own ramp runs a step under the canonical one across the middle of the range, which is why
 * the Android cut read denser and paler than the iOS one at the same palette: its two
 * workhorses were 14 sp and 16 sp Regular where the canonical ramp sets body and headline
 * a size up. The point sizes are Android's own — only the ROLE mapping is shared, and the
 * numbers themselves live in `App/Sources/Design/Theme.swift` for the iOS cut.
 *
 * Built on first composition rather than on class load, so a plain JVM unit test can read
 * the token tables in this file without a type ramp being raised behind it.
 */
private val SprossTypography: Typography by lazy {
    Typography().run {
        copy(
            displayLarge = displayLarge.rounded(),
            displayMedium = displayMedium.rounded(),
            // The emoji face of a Home card; a glyph, not a ramp entry.
            displaySmall = displaySmall.rounded(),
            // hero — screen titles.
            headlineLarge = headlineLarge.rounded(FontWeight.Bold, 34.sp),
            // The card headword, and the one slot with NO iOS counterpart: it sits between
            // `title` and `hero`, above the 22 pt the iOS card actually sets. Kept there
            // because the headword is the whole point of a card and Android has the room —
            // but it is this cut's own call, not a token copied across, and a word too wide
            // for its card steps down instead of breaking ([Headword]).
            headlineMedium = headlineMedium.rounded(FontWeight.Bold),
            // statValue — section headings and the numbers on a stat tile. Bold is the
            // point of it: a stat set Regular is the one thing on the screen that has to
            // be read at a glance and reads as body copy instead.
            headlineSmall = headlineSmall.rounded(FontWeight.Bold),
            // title — a card's own heading.
            titleLarge = titleLarge.rounded(FontWeight.Bold),
            // headline — the workhorse row label.
            titleMedium = titleMedium.rounded(FontWeight.SemiBold, 17.sp),
            // headline, minor.
            titleSmall = titleSmall.rounded(FontWeight.SemiBold, 15.sp),
            // body — real copy.
            bodyLarge = bodyLarge.rounded(size = 17.sp),
            // subheadline — supporting copy under a heading.
            bodyMedium = bodyMedium.rounded(size = 15.sp),
            // caption.
            bodySmall = bodySmall.rounded(FontWeight.Medium),
            // headline — button text, which M3 leaves at body weight.
            labelLarge = labelLarge.rounded(FontWeight.SemiBold, 15.sp),
            // badge.
            labelMedium = labelMedium.rounded(FontWeight.Bold, 13.sp),
            labelSmall = labelSmall.rounded(FontWeight.Medium),
        )
    }
}

private val LocalThemeColors = staticCompositionLocalOf { ThemeLight }

/**
 * The tokens M3 has no role for — amber, the article trio, the on-accent ink — read
 * through here, `MaterialTheme.colorScheme`'s sibling. Everything that DOES have a role
 * is read from the color scheme, so a component picks it up without being told.
 */
object Theme {
    val colors: ThemeColors
        @Composable @ReadOnlyComposable get() = LocalThemeColors.current

    val spacing = Spacing()
    val reserve = Reserve()
    val prompt = Prompt()

    /** Spacing, the same five steps and the same numbers the canonical table names. */
    class Spacing internal constructor() {
        val xs = 4.dp
        val sm = 8.dp
        val md = 12.dp
        val lg = 16.dp
        val xl = 24.dp
    }

    /**
     * Reserved heights, the same roles and the same numbers `Theme.reserve` names.
     *
     * A session prompt card reserves the height of its own tallest ROUTINE state, so nothing
     * below it moves as optional content comes and goes — vertical space is the scarce axis
     * (card, field, button and keyboard share one screen), and a button that slides under the
     * keyboard costs more than a card with air in it. The reveal is exempt: it grows the card
     * downward and reserves nothing.
     */
    class Reserve internal constructor() {
        /**
         * Drill prompt, shared by both drill faces: one big line of digits plus the
         * place-value pill (141.3 dp measured on the iOS cut), and the listening card's
         * caption plus replay glyph plus its once-per-run silent-switch line. A gap word
         * ("Ge l＿") is the exception that grows the card.
         */
        val drillCard = 144.dp

        /**
         * The floor a review card holds: one height whether the prompt is a word, a word under
         * an area label, or the replay glyph of a question asked by ear.
         */
        val reviewCard = 120.dp

        /**
         * A tappable tile's floor — a glyph over its label, at a size a thumb finds without
         * aiming. The letter drill's choices and the hub's entry chips are one size.
         */
        val tile = 72.dp
    }

    /**
     * The QUESTION itself, at the size its card can hold — the roles `Theme.typography.prompt` names,
     * cut for this platform.
     *
     * Fixed sp rather than a ramp slot: a prompt card reserves a height ([Theme.reserve.drillCard]),
     * and WHAT is asked picks the size — there is room for one numeral where there is none for a
     * whole sentence. A name needs no entry here: [Headword] already sizes one, and steps it
     * down to fit.
     */
    class Prompt internal constructor() {
        /** A numeral the whole card is about ("1 978", "14:35"). */
        val digits = 46.sp

        /** A picture standing where the name would: a flag that IS the question. */
        val glyph = 64.sp

        /** One word with a blank in it ("Ge l＿"). */
        val word = 30.sp

        /** A prompt made of words, laid out like one: wrapped over lines. */
        val sentence = 26.sp
    }
}

/**
 * The hue an ARTICLE wears: masculine blue, feminine berry, German's neuter green,
 * and nothing where the box cannot name a gender.
 *
 * Which article marks which gender is content, so [articleGender] answers it once for
 * every surface; only the three colors are this platform's, and they stay here. [lang]
 * settles the one form two languages share; a surface without it leaves that form neutral.
 */
fun ThemeColors.articleTint(article: String?, lang: Language? = null): Color? = when (articleGender(article, lang)) {
    Gender.Masculine -> der
    Gender.Feminine -> die
    Gender.Neuter -> das
    null -> null
}

@Composable
fun SprossTheme(content: @Composable () -> Unit) {
    // why: the system setting decides, which is why both columns exist — the iOS cut
    // reaches the same place through its dynamic color providers.
    val dark = isSystemInDarkTheme()
    CompositionLocalProvider(LocalThemeColors provides if (dark) ThemeDark else ThemeLight) {
        MaterialTheme(
            colorScheme = if (dark) SprossDark else SprossLight,
            shapes = SprossShapes,
            typography = SprossTypography,
            content = content,
        )
    }
}
